### PR TITLE
Add NoNullPropertyInitializationFixer

### PR DESCRIPTION
**Pull request by ntzm**
Implements #2806

Should this be configurable? i.e. if somebody wanted to have `= null;` on all of them?

Thanks for taking the time to review!

### CODE REVIEW

File: src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+            $equalsTokenIndex = $tokens->getNextMeaningfulToken($varTokenIndex);
+
+            if (!$tokens[$equalsTokenIndex]->equals('=')) {
+                continue;
+            }
+
+            $valueTokenIndex = $tokens->getNextMeaningfulToken($equalsTokenIndex);
+            $valueToken = $tokens[$valueTokenIndex];
+
+            if ($valueToken->isGivenKind(T_STRING) && $valueToken->getContent() === 'null') {
```
Comment (ntzm) [inline src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php]: I feel like there's a better way to detect `null`?

File: src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            $equalsTokenIndex = $tokens->getNextMeaningfulToken($varTokenIndex);
+
+            if (!$tokens[$equalsTokenIndex]->equals('=')) {
+                continue;
+            }
+
+            $valueTokenIndex = $tokens->getNextMeaningfulToken($equalsTokenIndex);
+            $valueToken = $tokens[$valueTokenIndex];
+
+            if ($valueToken->isNativeConstant() && strtolower($valueToken->getContent()) === 'null') {
```
Comment (julienfalque) [inline src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php]: Please use yoda style for the second condition.

File: src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+            $varTokenIndex = $tokens->getNextMeaningfulToken($index);
+
+            if (!$tokens[$varTokenIndex]->isGivenKind(T_VARIABLE)) {
+                continue;
+            }
+
+            $equalsTokenIndex = $tokens->getNextMeaningfulToken($varTokenIndex);
+
+            if (!$tokens[$equalsTokenIndex]->equals('=')) {
```
Comment (julienfalque) [inline src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php]: I think support for grouped properties is missing, e.g.:
```diff
-<?php class Foo { public $foo = null, $bar = null; }
+<?php class Foo { public $foo, $bar; }
```

**Comment by julienfalque** [comment]
> Should this be configurable? i.e. if somebody wanted to have `= null;` on all of them?

Not sure it would be widely used so maybe not worth it. I would suggest to focus on current behavior and see later for the configuration option.

File: src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            ]
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens)
+    {
+        foreach ($tokens as $index => $token) {
```
Comment (keradus) [inline src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php]: use regular `for` loop, and then you could make a jump of `$index = $valueTokenIndex;`, so you won't iterate tokens between those twice

File: src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php (hunk 2 above)
Comment (keradus) [inline src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php]: do not use `getContent` for comparision purposes.
use `->equals(..., false)` instead (false for case-insensitive)

File: tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            ],
+            [
+                '<?php class Foo { public $bar; protected $baz; }',
+                '<?php class Foo { public $bar = null; protected $baz = null; }',
+            ],
+            [
+                '<?php class Foo { public $bar = \'null\'; }',
+            ],
+            [
+                '<?php class Foo { public $bar; }',
```
Comment (keradus) [inline tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php]: not needed, this case is autogenerated from 1st case in this provider

File: tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+     * @dataProvider provideCases
+     */
+    public function testFix($expected, $input = null)
+    {
+        $this->doTest($expected, $input);
+    }
+
+    public function provideCases()
+    {
+        return [
```
Comment (keradus) [inline tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php]: please add case with property declared as final

**Comment by keradus** [comment]
I see no reason for configuring this fixer to be honset

File: tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php (hunk 6 above)
Comment (ntzm) [inline tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php]: It's not possible to declare a property as final in PHP

File: src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php (hunk 3 above)
Comment (ntzm) [inline src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php]: Working on this now

File: tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php (hunk 6 above)
Comment (keradus) [inline tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php]: too much java lately ;)

File: tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php [hunk 7]
```diff
@@ -20,6 +20,8 @@
  * @internal
  *
  * @covers \PhpCsFixer\Fixer\ClassNotation\NullPropertyDeclarationFixer
+ *
+ * @group new
```
Comment (keradus) [inline tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php]: i bet this is temporary. you could achieve same with `--filter` param without modifying the code itself

Comment (ntzm) [inline tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php]: Yep, forgot to remove! Thanks for the tip!

File: src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php (hunk 3 above)
Comment (ntzm) [inline src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php]: Having a bit of a nightmare getting this working lol, gonna give it a go when I have the chance in the week

File: tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+            [
+                '<?php class Foo { public $bar; protected $baz; }',
+                '<?php class Foo { public $bar = null; protected $baz = null; }',
+            ],
+            [
+                '<?php class Foo { public $bar = \'null\'; }',
+            ],
+            [
+                '<?php class Foo { public function bar() { return null; } }',
+            ],
```
Comment (SpacePossum) [inline tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php]: please add the tests from here as well;
https://github.com/SpacePossum/PHP-CS-Fixer/commit/756cd62c5b91383079de0491faadd5d7de110b72#diff-f58632716eb903df26e8da18f82e6016R42
(at least the one that cover a case not list here ;) )

File: src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php (hunk 3 above)
Comment (julienfalque) [inline src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php]: What is causing you trouble? Can we help you?

Comment (keradus) [inline src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php]: feel free to check out visibility fixer, which have to play with multiple properties as well

Comment (ntzm) [inline src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php]: Just trying to detect when it's a multiple properties, thanks @keradus I will check that out tonight!

**Comment by keradus** [comment]
ping @ntzm

**Comment by ntzm** [comment]
Sorry I haven't got the time to complete this, anyone is free to use all my code without crediting they want it to finish off the feature! Sorry about that

**Comment by julienfalque** [comment]
No worries. Too bad you didn't have time to complete this PR but I think that what you already did is great and should be kept so I'm reopening it. If you're ok with that, I'd like to push commits direcly here to add the missing part as soon as I can.

**Comment by ntzm** [comment]
Oh yeah that's fine sorry!!! Thanks for completing the rest of it 😆

File: tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php (hunk 8 above)
Comment (SpacePossum) [inline tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php]: missing anonymous class, traits etc. @ see https://github.com/SpacePossum/PHP-CS-Fixer/commit/756cd62c5b91383079de0491faadd5d7de110b72#diff-f58632716eb903df26e8da18f82e6016R42

Comment (julienfalque) [inline tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php]: Tests added.

Comment (SpacePossum) [inline tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php]: Still missing traits, multiple nested classes and multiple non nested classes in one file. Pease check the link for more test cases.

Comment (julienfalque) [inline tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php]: Added more tests. I think coverage is fine now.

Comment (SpacePossum) [inline tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php]: almost ;)  please update at least one test with diff. casing for `var` (like `VAR`) and one for scoping (like `PuliC`),
other missing is `public const $a = null` on PHP 7.1. and a test where the properties are defined after or in between methods.
And an un-godly comment one like;
```php
class A {
#
private#
#
$a#
#
=#
#
null#
,#
$b#
=#
null#
#
;#
#
}
```

After that I'm out of ideas

Comment (julienfalque) [inline tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php]: Added all cases except the one for `public const $a = null`, the assignment part is required when declaring a constant, even with `null`.

Comment (keradus) [inline tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php]: then please add `public const` case to prove it doesn't wrongly remove initial value

Comment (julienfalque) [inline tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php]: Of course, what was I thinking 😑. Test cases added.

File: src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+ * @author ntzm
+ */
+final class NullPropertyDeclarationFixer extends AbstractFixer
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function isCandidate(Tokens $tokens)
+    {
+        return $tokens->isAnyTokenKindsFound([T_PUBLIC, T_PROTECTED, T_PRIVATE, T_VAR]);
```
Comment (SpacePossum) [inline src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php]: T_CLASS or T_TRAIT must be found together with the current check

Comment (keradus) [inline src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php]: no need for, you cannot have `T_PUBLIC` etc without classy token, as it would be invalid code

Comment (SpacePossum) [inline src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php]: you can have T_PUBLIC in an interface which has no properties so is never a candidate

File: src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php [hunk 10]
```diff
@@ -66,10 +66,14 @@ protected function applyFix(\SplFileInfo $file, Tokens $tokens)
                 $index = $tokens->getNextMeaningfulToken($index);
 
                 if ($tokens[$index]->equals('=')) {
-                    $valueTokenIndex = $index = $tokens->getNextMeaningfulToken($index);
+                    $index = $tokens->getNextMeaningfulToken($index);
 
                     if ($tokens[$index]->equals([T_STRING, 'null'], false)) {
-                        $tokens->clearRange($varTokenIndex + 1, $valueTokenIndex);
+                        for ($i = $varTokenIndex + 1; $i <= $index; ++$i) {
+                            if (!$tokens[$i]->isComment() && !($tokens[$i]->isWhitespace() && false !== strpos($tokens[$i]->getContent(), "\n"))) {
```
Comment (julienfalque) [inline src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php]: I would swap the whitespace and the comment checks: whitespaces are more likely to be found here than comments, this should make this `if` slightly faster in such case.

Also, I would remove whitespaces even when they contain a newline (except when preceded by a single-line comment), WDYT?

Comment (SpacePossum) [inline src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php]: :+1: for the first (can you do it?),
second not so much, we have better fixers to deal with line breaks

Comment (julienfalque) [inline src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php]: Done.

File: tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+                '<?php class Foo { private $bar; }',
+                '<?php class Foo { private $bar = null; }',
+            ],
+            [
+                '<?php class Foo { var $bar; }',
+                '<?php class Foo { var $bar = null; }',
+            ],
+            [
+                '<?php class Foo { VAR $bar; }',
+                '<?php class Foo { VAR $bar = null; }',
```
Comment (keradus) [inline tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php]: we need a test with `= \null`, which won't be fixed with current implementation

File: src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+
+use PhpCsFixer\AbstractFixer;
+use PhpCsFixer\FixerDefinition\CodeSample;
+use PhpCsFixer\FixerDefinition\FixerDefinition;
+use PhpCsFixer\Tokenizer\Tokens;
+
+/**
+ * @author ntzm
+ */
+final class NullPropertyDeclarationFixer extends AbstractFixer
```
Comment (keradus) [inline src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php]: One day we normalized names to actually provide info what will be fixed version.
Here, we don't know

File: src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+
+namespace PhpCsFixer\Fixer\ClassNotation;
+
+use PhpCsFixer\AbstractFixer;
+use PhpCsFixer\FixerDefinition\CodeSample;
+use PhpCsFixer\FixerDefinition\FixerDefinition;
+use PhpCsFixer\Tokenizer\Tokens;
+
+/**
+ * @author ntzm
```
Comment (keradus) [inline src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php]: do you want to add you here guys ? @julienfalque @SpacePossum

File: src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+ */
+final class NullPropertyDeclarationFixer extends AbstractFixer
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function getDefinition()
+    {
+        return new FixerDefinition(
+            'Simplifies null class property declarations.',
```
Comment (keradus) [inline src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php]: simplifies... how ?

File: src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php (hunk 13 above)
Comment (julienfalque) [inline src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php]: I never add myself as `@author` so no, thanks :)

File: src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php (hunk 12 above)
Comment (julienfalque) [inline src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php]: `NoNullPropertyFixer`?

File: src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php (hunk 14 above)
Comment (julienfalque) [inline src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php]: What about _"Removes class properties initialization when the value is `null`."_?

File: src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php (hunk 12 above)
Comment (keradus) [inline src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php]: whole property would be removed? (even with that, code would work, so...)
`NoNullPropoertyInitializationFixer` ?

File: tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php (hunk 11 above)
Comment (julienfalque) [inline tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php]: Good catch!

File: src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php (hunk 14 above)
Comment (keradus) [inline src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php]: I bet it would follow decision from https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2822#discussion_r123598808 concern

File: src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php (hunk 12 above)
Comment (julienfalque) [inline src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php]: > even with that, code would work, so...

Haha, PHP...

👍 for `NoNullPropertyInitializationFixer`.

File: src/Fixer/ClassNotation/NoNullPropertyInitializationFixer.php [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+ */
+final class NoNullPropertyInitializationFixer extends AbstractFixer
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function getDefinition()
+    {
+        return new FixerDefinition(
+            'Removes class properties initialization when the value is `null`.',
```
Comment (keradus) [inline src/Fixer/ClassNotation/NoNullPropertyInitializationFixer.php]: following most (but not all) form of description, sth like:
> Properties MUST not be explicitly initialized with `null`.

Comment (julienfalque) [inline src/Fixer/ClassNotation/NoNullPropertyInitializationFixer.php]: Description updated.

File: src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php (hunk 12 above)
Comment (julienfalque) [inline src/Fixer/ClassNotation/NullPropertyDeclarationFixer.php]: Fixer renamed.

File: tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php (hunk 11 above)
Comment (julienfalque) [inline tests/Fixer/ClassNotation/NullPropertyDeclarationFixerTest.php]: Fixed.

**Comment by keradus** [comment]
Thank you @ntzm.

**Comment by SpacePossum** [comment]
thanks all authors :)  :+1:
